node/circuits.js

```diff
@@ -0,0 +1,154 @@
+        probation: options.probation,
+    };
+    self.shouldRequestOptions = {};
+}
+
+Circuits.prototype.getCircuit = function getCircuit(callerName, serviceName, endpointName) {
+    var self = this;
+    // TODO escape components
+    var circuitName = callerName + '::' + serviceName + '::' + endpointName;
+    var circuit = self.circuits[circuitName];
+    if (!circuit) {
+        circuit = new Circuit();
+        circuit.circuitName = circuitName;
+        circuit.callerName = callerName;
+        circuit.serviceName = serviceName;
+        circuit.endpointName = endpointName;
+        circuit.shouldRequestOptions = self.shouldRequestOptions;
+        circuit.stateOptions = self.stateOptions;
+        self.circuits[circuitName] = circuit;
+    }
+    return circuit;
+};
+
+Circuits.prototype.handleRequest = function handleRequest(req, buildRes, nextHandler) {
+    var self = this;
+    var callerName = req.headers.cn;
+    if (!callerName) {
+        return buildRes().sendError('BadRequest', 'All requests must have the cn (caller service name) header');
+    }
+    var serviceName = req.serviceName;
+    if (!serviceName) {
+        return buildRes().sendError('BadRequest', 'All requests must have a service name');
+    }
+    return req.withArg1(function withArg1(endpointName) {
+        var circuit = self.getCircuit(callerName, serviceName, endpointName);
+        return circuit.handleRequest(req, buildRes, nextHandler);
+    });
+};
+
+// Called upon membership change to collect services that the corresponding
+// egress node is no longer responsible for.
+Circuits.prototype.updateServices = function updateServices(serviceNames) {
+    var self = this;
+    var selfServiceNames = Object.keys(self.circuitsByServiceName);
+    for (var index = 0; index < selfServiceNames.length; index++) {
+        var selfServiceName = self.circuitsByServiceName[index];
+        if (serviceNames.indexOf(selfServiceNames) < 0) {
+            self.collectService(selfServiceName);
+        }
+    }
+};
+
+Circuits.prototype.collectService = function collectService(serviceName) {
+    var self = this;
+    var circuits = self.circuitsByServiceName[serviceName];
+    for (var index = 0; index < circuits.length; index++) {
+        delete self.circuits[circuits[index].circuitName];
+    }
+};
+
+function Circuit() {
+    var self = this;
+    self.circuitName = null;
+    self.callerName = null;
+    self.serviceName = null;
+    self.endpointName = null;
+    self.shouldRequestOptions = null;
+    self.stateOptions = null;
+    StateMachine.call(self);
+    self.setState(States.HealthyState);
+}
+
+Circuit.prototype.handleRequest = function handleRequest(req, buildRes, nextHandler) {
+    var self = this;
+    if (self.state.shouldRequest(req, self.shouldRequestOptions)) {
+        return self.monitorRequest(req, buildRes, nextHandler);
+    } else {
+        return buildRes().sendError('Declined', 'Service is not healthy');
```

Comment:
Thanks for asking. I reconsulted the :sparkles::book: protocol document :book::sparkles:, and I believe we want to avoid retries, which I believe actually means we should use UnexpectedError.